src/backend/distributed/planner/recursive_planning.c

```diff
@@ -1337,6 +1446,321 @@ NodeContainsSubqueryReferencingOuterQuery(Node *node)
+ * skip generating NULL's for non-required columns, and if we create (SELECT a, NULL, NULL FROM table)
+ * then this will be sent over network and NULL's also occupy some space. Instead of this we generate:
+ * (SELECT t.a, NULL, NULL FROM (SELECT a FROM table) t). The inner subquery will be recursively planned
+ * but the outer part will not be yet it will still have the NULL columns so that the query is correct.
+ */
+void
+ReplaceRTERelationWithRteSubquery(RangeTblEntry *rangeTableEntry,
+								  List *requiredAttrNumbers,
+								  RecursivePlanningContext *context)
+{
+	List *allTargetList = NIL;
+	Query *subquery = WrapRteRelationIntoSubquery(rangeTableEntry, requiredAttrNumbers,
+												  &allTargetList);
+	List *restrictionList =
+		GetRestrictInfoListForRelation(rangeTableEntry,
+									   context->plannerRestrictionContext);
+	List *copyRestrictionList = copyObject(restrictionList);
+	Expr *andedBoundExpressions = make_ands_explicit(copyRestrictionList);
+	subquery->jointree->quals = (Node *) andedBoundExpressions;
+	UpdateVarNosInNode(subquery, SINGLE_RTE_INDEX);
+
+	/* replace the function with the constructed subquery */
+	rangeTableEntry->rtekind = RTE_SUBQUERY;
+	rangeTableEntry->subquery = subquery;
+
+	/*
+	 * If the relation is inherited, it'll still be inherited as
+	 * we've copied it earlier. This is to prevent the newly created
+	 * subquery being treated as inherited.
+	 */
+	rangeTableEntry->inh = false;
+
+	if (IsLoggableLevel(DEBUG1))
+	{
+		char *relationAndAliasName = GetRelationNameAndAliasName(rangeTableEntry);
+		ereport(DEBUG1, (errmsg("Wrapping relation %s to a subquery",
+								relationAndAliasName)));
+	}
+
+	/* as we created the subquery, now forcefully recursively plan it */
+	bool recursivelyPlanned = RecursivelyPlanSubquery(subquery, context);
+	if (!recursivelyPlanned)
+	{
+		ereport(ERROR, (errmsg(
+							"unexpected state: query should have been recursively planned")));
+	}
+
+	Query *outerSubquery = CreateOuterSubquery(rangeTableEntry, allTargetList,
+											   requiredAttrNumbers);
+	rangeTableEntry->subquery = outerSubquery;
+}
+
+
+/*
+ * GetRelationNameAndAliasName returns the relname + alias name if
+ * alias name exists otherwise only the relname is returned.
+ */
+static char *
+GetRelationNameAndAliasName(RangeTblEntry *rangeTableEntry)
+{
+	StringInfo str = makeStringInfo();
+	appendStringInfo(str, "\"%s\"", get_rel_name(rangeTableEntry->relid));
+
+	char *aliasName = NULL;
+	if (rangeTableEntry->alias)
+	{
+		aliasName = rangeTableEntry->alias->aliasname;
+	}
+
+	if (aliasName)
+	{
+		appendStringInfo(str, " \"%s\"", aliasName);
+	}
+	return str->data;
+}
+
+
+/*
+ * CreateOuterSubquery creates outer subquery which contains
+ * the given range table entry in its rtable.
+ */
+static Query *
+CreateOuterSubquery(RangeTblEntry *rangeTableEntry, List *allTargetList,
+					List *requiredAttrNumbers)
+{
+	MakeVarAttNosSequential(allTargetList);
```

Comment:
could we do this when we generate allTargetList?

Reply:
we should be able to.

`WrapRteRelationIntoSubquery` is a bit messy, needs some refactoring, but I will probably just call `MakeVarAttNosSequential` there